### DISCUSSION TITLE
Journal Templates

### DISCUSSION THREAD

**Discussion by sez11a** [lervag/wiki.vim#349] [Q&A] [answered]
Hi! 

I'm working on converting from [Logseq](https://logseq.com) over to wiki.vim, due to my desire to keep everything in Markdown and human-readable. I read with interest the documentation on templates, but I'm not sure I get how to make them work. 

I'm using Neovim with the Lazy package manager. I converted the setup for journals in the docs over to Lua: 

```lua
return {
  {
  'lervag/wiki.vim',
    config = function()
      vim.g.wiki_journal = {
        name = 'journal',
        root = '',
        frequency = 'daily',
        date_format = {
          daily = '%Y/%m/%d',
          weekly = '%Y/week_%V',
          monthly = '%Y/%m/summary',
        },
      }
    end
  },
  {
  'lervag/lists.vim',
  }
}
```
Using this setup, I want to make a template for both the monthly and the daily journal. The docs mention a file called `.template.md`, but it's not clear where to put it, and it's not clear what to do if you want multiple templates (i.e., one for monthly and one for daily). 

I am a tech writer, so I would be happy to contribute documentation back in return for some help in setting this up. 

Thank you! 

--Rich

**Comment by lervag** [comment]
> I'm using Neovim with the Lazy package manager. I converted the setup for journals in the docs over to Lua: …

I assume this is a lazy.nvim configuration. First, you should use the `init` function to set options. The config block ends up setting the option too late to be fully effective. Also, you don't need to specify things that you don't change from default values. Thus, this should work:

```lua
return {
  {
    'lervag/wiki.vim',
    init = function()
      vim.g.wiki_journal = {
        date_format = {
          daily = '%Y/%m/%d',
          weekly = '%Y/week_%V',
          monthly = '%Y/%m/summary',
        },
      }
    end
  },
  {
    'lervag/lists.vim',
  }
}
```

Notice, though, I would suggest you also define the `vim.g.wiki_root` option.

Now, the templates are slightly complicated. First, there are builtin templates for the monthly and weekly frequencies that are made to summarize the daily and weekly entries, respectively. These builtin templates will copy contents from the more frequent steps. These templates are really built around my own personal preference/need and may not be what you want.

Let's consider the daily template, or for that matter, just any template you want for a "general" wiki page: you want to use the `vim.g.wiki_templates` option. As an example:

```lua
return {
  {
    'lervag/wiki.vim',
    init = function()
      vim.g.wiki_root = "/home/USER/wiki/.templates"
      vim.g.wiki_templates = {
        {
          match_re = "\d\d\d\d\/\d\d\/\d\d",
          source_filename = "/home/USER/wiki/.templates/daily.md"
        },
      },
    end
  },
```

where the daily template `daily.md` could look something like this:

```md
# {name}

# Plans

# Meetings
```

The relevant parts of the current docs are `:help g:wiki_templates` and `:help wiki-templates-format`. The template mechanism is quite flexible. If you are somewhat into programming, you can use a function both for the matcher and the source in the `wiki_templates` list.

Notice that any wiki pages will only have _one_ template. Thus, if you define a template that matches any weekly or monthly pages, then the builtin tempolates will be deactivated.

> I am a tech writer, so I would be happy to contribute documentation back in return for some help in setting this up.

If my help motivates you to suggest improvements to the docs I'll be very happy! :)

**Comment by sez11a** [comment]
Okay; here's what I've got so far, but it's not working: 

```lua
return {
  {
  'lervag/wiki.vim',
    init = function()
      vim.g.wiki_journal = {
        name = 'journal',
        root = '',
        frequency = 'daily',
        date_format = {
          daily = '%Y/%m/%d',
          weekly = '%Y/week_%V',
          monthly = '%Y/%m/summary',
        },
      }
      vim.g.wiki_templates = {
        { 
          match_re = "%d%d%d%d/%d%d/%d%d",
          source_filename = "~/.VimStar/wiki/templates/daily.md"
        },
      }
    end
  },
  {
  'lervag/lists.vim',
  }
}
```

As you can see, I changed the `match_re` variable because I was getting an Invalid Escape Sequence error. I'm no Lua expert, so I looked it up. According to [this page](https://www.lua.org/pil/20.2.html) you can do matches for digits with %, not \\. 

I have my template in the proper location, and I set the wiki location in another file, but I still get a blank file when I use <leader>w<leader>w to open the journal.

**Comment by lervag** [comment]
> Okay; here's what I've got so far, but it's not working: …
> 
> As you can see, I changed the `match_re` variable because I was getting an Invalid Escape Sequence error.

It is _not_ a Lua pattern, it is a Vim regex. So my version was correct, except it should be `[[...]]` instead of `"..."`.

Also, not sure why you still include the keys you are not changing from their default values...

So, this should work:

```lua
return {
  {
    'lervag/wiki.vim',
    init = function()
      vim.g.wiki_journal = {
        date_format = {
          daily = '%Y/%m/%d',
          weekly = '%Y/week_%V',
          monthly = '%Y/%m/summary',
        },
      }
      vim.g.wiki_templates = {
        {
          match_re = [[\d\d\d\d\/\d\d\/\d\d]],
          source_filename = "/home/USER/wiki/.templates/daily.md"
        },
      },
    end
  },
  {
    'lervag/lists.vim',
  }
}
```

**Comment by sez11a** [comment]
Thank you for the correct syntax. For whatever reason, though, I'm still getting a blank file instead of my template when I use the code above. 

Also, as far as I can tell, I'm not using the defaults. From your code (which is in my `~/.local/share/nvim/lazy/wiki.vim/plugin/wiki.vim`), you have these defaults: 

```vim
call wiki#init#option('wiki_journal', {
      \ 'name' : 'journal',
      \ 'root' : '',
      \ 'frequency' : 'daily',
      \ 'date_format' : {
      \   'daily' : '%Y-%m-%d',
      \   'weekly' : '%Y_w%V',
      \   'monthly' : '%Y_m%m',
      \ },
      \})
```
This resulted in a more flat structure for the journal. I preferred what you had in the documentation, which breaks the journal out into year/month directories and then uses the day as the title for the file. That's why I'm specifying that in my configuration.

**Comment by lervag** [comment]
> Thank you for the correct syntax. For whatever reason, though, I'm still getting a blank file instead of my template when I use the code above.

To start, you could try to use a more including regex. E.g., try `.*`, which should match _anything_. If that works, then we know the problem is with the regex - if it doesn't work, then the problem is with the template. E.g., perhaps the path you are using is wrong.

> Also, as far as I can tell, I'm not using the defaults.

What I mean is that you don't need to specify the keys that you don't change. That is, you are only changing the `date_format` part, so you don't need to specify `name`, `root` and `frequency`.

----

Btw., I'm not used to the discussion feature on Github. It seems we should have used the comment thread to my initial reply instead of adding new comments. Not really important, just thought I'd mention it.

**Reply by sez11a** [reply]
Oh, okay. Got it. 

Well, I haven't gotten anything to match except for `.*`, so at least I know my template can be found. I've tried

```vim
[[....\/..\/..]]
[[.*\d\d\d\d\/\d\d\/\d\d]]
[[****\/**\/**]] (invalid syntax)
[[*\/*\/*]]
[[0-9][0-9][0-9][0-9]\/[0-9][0-9]\/[0-9][0-9]]
```
and a bunch of others I don't remember. I'll keep goofing around with it. I'm also no regex expert, so if you have any ideas, I'm all ears. :-)

**Reply by sez11a** [reply]
BTW, the path is being created as expected: 

![image](https://github.com/lervag/wiki.vim/assets/515497/385ac805-78b0-48d9-9fb4-6f8b8fbd3ab7)

**Reply by lervag** [reply, marked as answer]
Ah, sorry, I just remembered - the `match_re` is only matched against the _name_, not the path. So, you should do this:

```lua
vim.g.wiki_journal = {
  date_format = {
    daily = '%Y/%m/%d',
    weekly = '%Y/week_%V',
    monthly = '%Y/%m/summary',
  },
}
vim.g.wiki_templates = {
  {
    match_func = function(context)
      return context.path_wiki:find "%d%d%d%d/%d%d/%d%d"
    end,
    source_filename = "/path/to/templates/daily.md"
  },
},
```

Notice, here we use Lua patterns, since we are matching with a Lua function.

**Reply by sez11a** [reply]
Yes! That did it! Awesome; thank you! 

I will write this up (maybe tomorrow) and send you a pull request for the docs.

**Reply by lervag** [reply]
Great, glad to hear it!
